add layer for npm packages

Toxicable opened the pull request:
Adds support for `NodeRuntimeDepsInfo` and `NpmPackageInfo`
This ensures that npm deps are placed in their own layer, improving caching and rebuilds
Also puts nodejs and nodejs args so that it can be cached between instances of nodejs_image targets.

based off work started here https://github.com/bazelbuild/rules_docker/pull/487/files

Toxicable commented:
/assign @smukherj1

Toxicable commented:
Just tested these changes on our internal repository and looks like it's working as expected

```
62bba9724dca: Loading layer [==>]  10.24kB/10.24kB
02fa2ebc0496: Loading layer [==>]  41.12MB/41.12MB
32e8dde897ae: Loading layer [==>]  10.24kB/10.24kB
078828ab7718: Loading layer [==>]  213.1MB/213.1MB
f432715cf9ce: Loading layer [==>]  1.444MB/1.444MB
84ff92691f90: Loading layer [==>]  10.24kB/10.24kB
```

Toxicable commented on container/image_test.py:
```diff
@@ -760,6 +760,185 @@ def test_war_image(self):
+        './app',
+        './app/testdata',
+        './app/testdata/nodejs_image_binary.runfiles',
+        './app/testdata/nodejs_image_binary.runfiles/nodejs_darwin_amd64',
```
how do we make this test platform agnostic?

gregmagolan replied:
or should this be platform specific based on the container OS?

nlopezgi replied:
this should be platform specific based on the container OS used in tests.

Toxicable replied:
adjusted it to `nodejs_linux_amd64`

gregmagolan commented on nodejs/image.bzl:
```diff
@@ -88,6 +86,43 @@ _dep_layer = rule(
+    for pkg in dep[NodeRuntimeDepsInfo].pkgs:
+        if NpmPackageInfo not in pkg:
+            depsets.append(pkg.files)
+    print(depsets)
```
nit: debug print

gregmagolan commented on container/image_test.py:
```diff
@@ -760,6 +760,185 @@ def test_war_image(self):
+        '/app/testdata/nodejs_image_binary', 
+        '/app/testdata/nodejs_image_binary.runfiles', 
+        '/app/testdata/nodejs_image_binary.runfiles/io_bazel_rules_docker', 
+        '/app/testdata/nodejs_image_binary.runfiles/io_bazel_rules_docker/external'
```
If you add `--build noexternal_legacy_runfiles` to `.bazelrc` the `/app/testdata/nodejs_image_binary.runfiles/io_bazel_rules_docker/external` should away which would be preferred for this test and should be the default in Bazel at some point. I was hoping they would flip that for 1.0 but they didn't yet.

Toxicable replied:
hmmm I cant see any trace of that flag throughout the repository, ill add a issue then put a TODO to remove these entries since im sure it'll impact other containers and their tests

Toxicable replied:
https://github.com/bazelbuild/rules_docker/issues/1350

gregmagolan replied:
👍

gregmagolan commented on nodejs/image.bzl:
```diff
@@ -115,23 +148,36 @@ def nodejs_image(
+    nodejs_layers = [
+        # Put the Node binary into its own layers.
+        "@nodejs//:node",
+        "@nodejs//:node_files",
```
FYI. The `@nodejs` repo will control which node platform is included.
Its an alias repo to the the host system platform. Either `@nodejs_darwin_amd64`, `@nodejs_linux_amd64` or `@nodejs_windows_amd64`.

Toxicable replied:
awesome so this hsould be fine with targetting different platforms then

gregmagolan commented:
Nice. I'm glad to see the `node_modules` attribute removed as I'd like to remove it from rules_nodejs but this repo was blocking that.

I'd like to dig into this a bit further and understand it in detail next week as this was something the @Globegitter started working on here https://github.com/bazelbuild/rules_nodejs/pull/863 but that was before we had `NodeRuntimeDepsInfo`. Also, for the platform choice to put in the docker image I think that should be controlled with the `--platform` Bazel CLI option.

There's an example usage in https://github.com/bazelbuild/rules_nodejs/blob/master/examples/nestjs/src/BUILD.bazel#L48:

```
# bazel build --platforms=@build_bazel_rules_nodejs//toolchains/node:linux_amd64 //src:docker
nodejs_image(
    name = "docker",
    data = [
        ":app",
    ],
    entry_point = ":main.ts",
    # TODO: switch to fine grained deps -- see bazelbuild/rules_docker#1000
    # and add test for docker image
    node_modules = "@npm//:node_modules",
)
```

nlopezgi commented on nodejs/image.bzl:
```diff
@@ -88,6 +86,42 @@ _dep_layer = rule(
     implementation = _dep_layer_impl,
 )
 
+def _npm_deps_runfiles(dep):
```
add brief doc

Toxicable replied:
added a doc, does that make sense what it does?

nlopezgi commented on nodejs/image.bzl:
```diff
@@ -88,6 +86,42 @@ _dep_layer = rule(
+    return depset(transitive = depsets)
+
+def _app_layer_impl(ctx):
+    return lang_image.implementation(ctx, runfiles = _runfiles, emptyfiles = _emptyfiles)
```
where is the entrypoint now being set since you're not calling app_layer any longer?

Toxicable replied:
Not sure what you're referring to as the entrypoint, do you mean the `entry_point` that exists on a `nodejs_binary`?
Because that gets collected via the `DeafultInfo` of the `nodejs_binary` on `_app_layer_impl`.
Note the `./app/testdata/nodejs_image_binary.runfiles/io_bazel_rules_docker/testdata/nodejs_image_binary.sh` in the above test, that's the script called to start the process inside the container.

nlopezgi replied:
Let me expand a little on my question.
Currently, if you run a nodehs_image target, the entrypoint to the container is set [here](https://github.com/bazelbuild/rules_docker/blob/master/lang/image.bzl#L208). When I print its value for` //container/nodejs:nodejs_image` the entrypoint to the container is set to `"/app/tests/container/nodejs/nodejs_image.binary"`
Now in your change, you are replacing the call to `app_layer` in nodejs/image/bzl for a call to `_app_layer`. the rule `_app_layer `is implemented by `_app_layer_impl` which in turn is calling `lang_image.implementation(ctx, runfiles = _runfiles, emptyfiles = _emptyfiles)`. This call to `lang_image.implementation` is not setting any value for the entrypoint of the container. So where is the value for the entrypoint for the nodejs_image now being set?

Toxicable replied:
Ah I think I understand, initally I though we'd need a app specific layer there. But turns out it's not needed. I've switched back to use the `app_layer` from `//lang:image.bzl`.
Which should fix up the entrypoint issues

Toxicable commented on container/image_test.py:
```diff
@@ -760,6 +760,189 @@ def test_war_image(self):
+        './app/testdata/nodejs_image_binary.runfiles/io_bazel_rules_docker/testdata/nodejs_image_lib.d.ts', 
+        './app/testdata/nodejs_image_binary.runfiles/build_bazel_rules_nodejs', 
+        './app/testdata/nodejs_image_binary.runfiles/build_bazel_rules_nodejs/internal', 
+        './app/testdata/nodejs_image_binary.runfiles/build_bazel_rules_nodejs/internal/linker', 
```
I'd like to put the files here from rules_nodejs into their own layer. Making this top layer only actual source files / generates files under the `io_bazel_rules_docker` workspace.  
But even with how it looks right now is a big improvement

Toxicable replied:
Will adress this one another time

nlopezgi commented:
Added more details to my question. Do let me know when this is ready for another review (i.e., make sure buildkite tests are passing)

Toxicable commented:
@nlopezgi Should be good for review again, addressed those points and CI is green.

nlopezgi commented:
/gcbrun

nlopezgi commented:
The nodejs image integration test is failing.
https://github.com/bazelbuild/rules_docker/blob/master/testing/e2e/nodejs_image.sh
```
This is the error message:
INFO: Build completed successfully, 55 total actions
Error: Cannot find module './nodejs_image_lib'. Please verify that the package.json has a valid "main" entry
    at Function.module.constructor._resolveFilename (tests/container/nodejs/nodejs_image.binary_require_patch.js:482:17)
    at Function.Module._load (internal/modules/cjs/loader.js:687:27)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (testdata/nodejs_image.js:2:13)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Object.<anonymous> (tests/container/nodejs/nodejs_image.binary_loader.js:32:24)
+ EXPECT_CONTAINS 'Loaded image ID: sha256:e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059
Tagging e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059 as bazel/tests/container/nodejs:nodejs_image' 'Hello World!'
+ local 'complete=Loaded image ID: sha256:e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059
Tagging e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059 as bazel/tests/container/nodejs:nodejs_image'
+ local 'substring=Hello World!'
+ local 'message=Expected '\''Hello World!'\'' not found in '\''Loaded image ID: sha256:e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059
Tagging e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059 as bazel/tests/container/nodejs:nodejs_image'\'''
+ echo Checking 'Loaded image ID: sha256:e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059
Tagging e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059 as bazel/tests/container/nodejs:nodejs_image' contains 'Hello World!'
+ CONTAINS 'Loaded image ID: sha256:e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059
Tagging e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059 as bazel/tests/container/nodejs:nodejs_image' 'Hello World!'
+ local 'complete=Loaded image ID: sha256:e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059
Tagging e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059 as bazel/tests/container/nodejs:nodejs_image'
+ local 'substring=Hello World!'
+ grep -Fsq -- 'Hello World!'
Checking Loaded image ID: sha256:e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059
Tagging e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059 as bazel/tests/container/nodejs:nodejs_image contains Hello World!
FAILURE: Expected 'Hello World!' not found in 'Loaded image ID: sha256:e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059
Tagging e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059 as bazel/tests/container/nodejs:nodejs_image'
+ echo 'Loaded image ID: sha256:e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059
Tagging e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059 as bazel/tests/container/nodejs:nodejs_image'
+ fail 'Expected '\''Hello World!'\'' not found in '\''Loaded image ID: sha256:e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059
Tagging e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059 as bazel/tests/container/nodejs:nodejs_image'\'''
+ echo 'FAILURE: Expected '\''Hello World!'\'' not found in '\''Loaded image ID: sha256:e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059
Tagging e7b6705db119ddb2f990e3fc6c79caa1c0f7729035fbb05583693493785f3059 as bazel/tests/container/nodejs:nodejs_image'\'''
+ exit 1
```
Failure corresponds to [this](https://github.com/bazelbuild/rules_docker/blob/master/testing/e2e/nodejs_image.sh#L26) line. You can probably repro it locally by running  `bazel run tests/container/nodejs:nodejs_image` or running the nodejs_image.sh script directly.

Toxicable commented:
I ran `bazel test tests/container/nodejs/...` and all passes locally for me now. I adjusted the testdata to include the extra lib and fixed up those build targets to correctly include the lib outputs.

Toxicable commented:
/gcburn

nlopezgi commented:
/gcbrun
